src/Compiler/Facilities/BuildGraph.fs

```diff
@@ -225,87 +36,86 @@ module GraphNode =
 #endif
         | None -> ()
 
+type ComputationState<'T> =
+    | Initial of Async<'T>
+    // TaskCompletionSource holds the result after completion.
+    | Requested of TaskCompletionSource<'T> * CancellationTokenSource
+
 [<Sealed>]
-type GraphNode<'T> private (computation: NodeCode<'T>, cachedResult: ValueOption<'T>, cachedResultNode: NodeCode<'T>) =
+type GraphNode<'T> private (initialState: ComputationState<'T>) =
 
-    let mutable computation = computation
     let mutable requestCount = 0
+    let mutable state = initialState
+
+    let startNew computation =
+        function
+        | Requested(tcs, cts) ->
+            // Must be started in the thread pool for performance reasons.
+            // StartImmediate / StartWithContinuations gives 2x worse performance, causing CI to timeout.
+            Async.Start(
+                async {
+                    try
+                        Thread.CurrentThread.CurrentUICulture <- GraphNode.culture
+                        let! result = computation
+                        tcs.SetResult result
+                    with
+                    | :? OperationCanceledException ->
+                        tcs.SetCanceled()
+                        // Make ready for restart.
+                        state <- Initial computation
+                    | ex ->
+                        tcs.SetException ex
+                        // Make ready for restart.
+                        state <- Initial computation
+                },
+                cts.Token
+            )
+        | _ -> ()
 
-    let mutable cachedResult = cachedResult
-    let mutable cachedResultNode: NodeCode<'T> = cachedResultNode
-
-    let isCachedResultNodeNotNull () =
-        not (obj.ReferenceEquals(cachedResultNode, null))
-
-    let semaphore = new SemaphoreSlim(1, 1)
-
-    member _.GetOrComputeValue() =
-        // fast path
-        if isCachedResultNodeNotNull () then
-            cachedResultNode
-        else
-            node {
-                Interlocked.Increment(&requestCount) |> ignore
-
-                try
-                    let! ct = NodeCode.CancellationToken
+    member this.GetOrComputeValue() =
```

Comment:
Perhaps you pruned this code a little too much? Why not start with keeping it as-is just without `node`?